gateway/ln-gateway/src/lightning/ldk.rs

```diff
@@ -0,0 +1,501 @@
+                        .to_string(),
+            }),
+        }?;
+
+        Ok((route_htlc_stream, Arc::new(*self)))
+    }
+
+    async fn complete_htlc(
+        &self,
+        htlc: InterceptHtlcResponse,
+    ) -> Result<EmptyResponse, LightningRpcError> {
+        let InterceptHtlcResponse {
+            action,
+            payment_hash,
+            incoming_chan_id: _,
+            htlc_id: _,
+        } = htlc;
+
+        let ph = PaymentHash(payment_hash.clone().try_into().unwrap());
+
+        // TODO: Get the actual amount from the LDK node. Probably makes the
+        // most sense to pipe it through the `InterceptHtlcResponse` struct.
+        // This value is only used by `ldk-node` to ensure that the amount
+        // claimed isn't less than the amount expected, but we've already
+        // verified that the amount is correct when we intercepted the HTLC.
+        let claimable_amount_msat = 999_999_999_999_999;
+
+        let ph_hex_str = hex::encode(payment_hash);
+
+        match action {
+            Some(Action::Settle(Settle { preimage })) => {
+                self.node
+                    .bolt11_payment()
+                    .claim_for_hash(
+                        ph,
+                        claimable_amount_msat,
+                        PaymentPreimage(preimage.try_into().unwrap()),
+                    )
+                    .map_err(|_| LightningRpcError::FailedToCompleteHtlc {
+                        failure_reason: format!(
+                            "Failed to claim LDK payment with hash {ph_hex_str}"
+                        ),
+                    })?;
+            }
+            _ => {
+                self.node.bolt11_payment().fail_for_hash(ph).map_err(|_| {
+                    LightningRpcError::FailedToCompleteHtlc {
+                        failure_reason: format!(
+                            "Failed to unwind LDK payment with hash {ph_hex_str}"
+                        ),
+                    }
+                })?;
+            }
+        };
+
+        return Ok(EmptyResponse {});
+    }
+
+    async fn create_invoice(
+        &self,
+        create_invoice_request: CreateInvoiceRequest,
+    ) -> Result<CreateInvoiceResponse, LightningRpcError> {
+        let payment_hash = PaymentHash(create_invoice_request.payment_hash.try_into().map_err(
+            |_| LightningRpcError::FailedToGetInvoice {
+                failure_reason: "Failed to convert Vec<u8> to [u8; 32] (this probably means that LDK received an invalid payment hash)".to_string(),
+            },
+        )?);
+
+        let invoice = self
+            .node
+            .bolt11_payment()
+            .receive_for_hash(
+                create_invoice_request.amount_msat,
+                if let Some(Description::Direct(description)) = &create_invoice_request.description
+                {
+                    description
+                } else {
+                    ""
```

Comment:
does ldk node support setting the description hash in `receive_for_hash`?

Reply:
I don't think so. Drilling down into the function calls here, we end up at this:

https://github.com/lightningdevkit/rust-lightning/blob/f48a273211d5235e1ff335b2c25f8aacf3642f24/lightning-invoice/src/utils.rs#L431-L453

Where `_create_invoice_from_channelmanager_and_duration_since_epoch()` accepts a `Bolt11InvoiceDescription‎` (which can be a direct description or a hash) but is hardcoded to always be a direct description.

Reply:
This would probably be good to make an issue upstream

Reply:
Created https://github.com/lightningdevkit/ldk-node/issues/325